[Pytorch] Existing MHA: do in-place division on q

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #73249
* #73219

q is never used again after this line, may as well avoid a copy.

Differential Revision: [D34400111](https://our.internmc.facebook.com/intern/diff/D34400111/)

swolchok committed:
[Pytorch] Existing MHA: do in-place division on q

q is never used again after this line, may as well avoid a copy.

Differential Revision: [D34400111](https://our.internmc.facebook.com/intern/diff/D34400111/)

[ghstack-poisoned]

albanD commented on torch/nn/functional.py:
```diff
@@ -5030,7 +5030,7 @@ def _scaled_dot_product_attention(
     """
     B, Nt, E = q.shape
-    q = q / math.sqrt(E)
+    q /= math.sqrt(E)
```
I would put a comment in the doc or next to the call site mentioning that. After calling this function, the q that was given should NOT be re-used anymore.

swolchok committed:
Update on "[Pytorch] Existing MHA: do in-place division on q"


q is never used again after this line, may as well avoid a copy.

Differential Revision: [D34400111](https://our.internmc.facebook.com/intern/diff/D34400111/)

[ghstack-poisoned]

swolchok committed:
Update on "[Pytorch] Existing MHA: do in-place division on q"


q is never used again after this line, may as well avoid a copy.

Differential Revision: [D34400111](https://our.internmc.facebook.com/intern/diff/D34400111/)

[ghstack-poisoned]

swolchok committed:
Update on "[Pytorch] Existing MHA: do in-place division on q"


q is never used again after this line, may as well avoid a copy.

Differential Revision: [D34400111](https://our.internmc.facebook.com/intern/diff/D34400111/)

[ghstack-poisoned]

swolchok commented:
not compatible with autograd & we have fast path now
